spyder/plugins/editor/widgets/completion.py

```diff
@@ -0,0 +1,401 @@
+                item.setData(Qt.UserRole, completion_label)
+                item.setText(completion_text)
+
+            if self.check_can_complete(
+                    completion_label, filter_text):
+                self.addItem(item)
+
+        if self.count() > 0:
+            self.setCurrentRow(0)
+            self.scrollTo(self.currentIndex(),
+                          QAbstractItemView.PositionAtTop)
+        else:
+            self.hide()
+
+    def get_html_item_representation(self, item_completion, item_type,
+                                     height=DEFAULT_COMPLETION_ITEM_HEIGHT,
+                                     width=DEFAULT_COMPLETION_ITEM_WIDTH):
+        """Get HTML representation of and item."""
+
+        return COMPLETION_ITEM_TEMPLATE.format(completion=item_completion,
+                                               type=item_type,
+                                               color=ima.MAIN_FG_COLOR,
+                                               height=height,
+                                               width=width)
+
+    def hide(self):
+        """Hide the widget."""
+        self.completion_position = None
+        self.completion_list = None
+        self.clear()
+        QToolTip.hideText()
+        QListWidget.hide(self)
+        self.textedit.setFocus()
+
+    def keyPressEvent(self, event):
+        """Process keypress."""
+        text, key = event.text(), event.key()
+        alt = event.modifiers() & Qt.AltModifier
+        shift = event.modifiers() & Qt.ShiftModifier
+        ctrl = event.modifiers() & Qt.ControlModifier
+        modifier = shift or ctrl or alt
+        if key in (Qt.Key_Return, Qt.Key_Enter, Qt.Key_Tab):
+            # Check that what was selected can be selected,
+            # otherwise timing issues
+            if self.up_to_date():
+                self.item_selected()
+            else:
+                self.hide()
+                self.textedit.keyPressEvent(event)
+        elif key == Qt.Key_Escape:
+            self.hide()
+        elif key in (Qt.Key_Left, Qt.Key_Right) or text in ('.', ':'):
+            self.hide()
+            self.textedit.keyPressEvent(event)
+        elif key in (Qt.Key_Up, Qt.Key_Down, Qt.Key_PageUp, Qt.Key_PageDown,
+                     Qt.Key_Home, Qt.Key_End,
+                     Qt.Key_CapsLock) and not modifier:
+            if key == Qt.Key_Up and self.currentRow() == 0:
+                self.setCurrentRow(self.count() - 1)
+            elif key == Qt.Key_Down and self.currentRow() == self.count()-1:
+                self.setCurrentRow(0)
+            else:
+                QListWidget.keyPressEvent(self, event)
+        elif len(text) or key == Qt.Key_Backspace:
+            self.textedit.keyPressEvent(event)
+            self.update_current()
+        elif modifier:
+            self.textedit.keyPressEvent(event)
+        else:
+            self.hide()
+            QListWidget.keyPressEvent(self, event)
+
+    def up_to_date(self):
```

Comment:
We should improve this name, since it returns a boolean it should be more like

`check_up_to_date` or `is_up_to_date`